Nav 131 extend query config for raptor

clukas1 opened the pull request:
@munterfi this is part one of my query config issue. I thought I'll initiate the review process early to unlock the service for further extension (less merge conflicts).

So far I've extended everything on the app / spring endpoint side to handle extra query args as I would expect. Down to creating the service QueryConfig object. In a separate branch, I will use this QueryConfig object to build a raptor query config object and mask my stop times accordingly + tests.

A further issue I will tackle is creating multiple versions of my zürich trams gtfs schedule where I add accesibility/bike information for testing.

clukas1 commented:
@munterfi: PR is finished for review and potential merger. Only thing one could still consider with a slight potential for adding too much complexity is considering the accessibility of stops in the routing process as well. This would require reworking the raptor builder and router a bit so that stops would also have a flag for accessibility.

munterfi reviewed (changes requested):
Thanks @clukas1, nice work!

The overall concept looks solid and is close to being ready for merge. I have a few minor concerns that I'd like to discuss:

- `RoutingInfo` vs. `ScheduleInfo`: Not consistent where this should live (`ScheduleInfomrationService` vs. `ConnectionRoutingService`) between the layers `service` and `app`.
- Why is there and option to have no travel modes? Do you know sources other than GTFS that are not providing this? In order to be generic we could keep it. But in our case this will always be `supportsTravelModes=true`.
- Configuration specific raptors and trip masks: Could you profile and benchmark to see how this is going to impact our memory footprint and performance?

clukas1 commented on src/main/java/ch/naviqore/app/dto/TimeType.java:
```diff
@@ -1,6 +0,0 @@
-package ch.naviqore.app.dto;
-
-public enum TimeType {
```
Mainly laziness of adding map() to all conversions but after consulting with chatgpt and getting a clear yes that I should separate dtos and service enums, I've reintroduced it: 71b7779

clukas1 commented on src/main/java/ch/naviqore/raptor/router/StopTimeProvider.java:
```diff
@@ -48,13 +49,20 @@ class StopTimeProvider {
+    private String getCacheKeyForStopTimes(LocalDate date, QueryConfig queryConfig) {
         String serviceId = tripMaskProvider.getServiceIdForDate(date);
-        return stopTimeCache.computeIfAbsent(serviceId, () -> createStopTimesForDate(date));
+        String queryConfigKey = String.format("%b-%b-%s", queryConfig.isWheelchairAccessible(),
```
99962b7

clukas1 commented on src/main/java/ch/naviqore/service/gtfs/raptor/convert/GtfsTripMaskProvider.java:
```diff
@@ -28,6 +37,38 @@ public GtfsTripMaskProvider(GtfsSchedule schedule, int cacheSize, EvictionCache.
     }
 
+    private static EnumSet<DefaultRouteType> mapToRouteTypes(EnumSet<TravelMode> travelModes) {
+        if (travelModes == null || travelModes.isEmpty()) {
```
d7c1475

clukas1 commented on src/main/java/ch/naviqore/service/gtfs/raptor/convert/GtfsTripMaskProvider.java:
```diff
@@ -38,22 +79,60 @@ public String getServiceIdForDate(LocalDate date) {
+
             boolean[] tripMask = new boolean[tripIds.length];
+
+            // only check route type if there is a non-default travel mode filter
```
d7c1475

clukas1 commented on src/main/java/ch/naviqore/raptor/router/RaptorTripMaskProvider.java:
```diff
@@ -32,14 +34,15 @@ public interface RaptorTripMaskProvider {
+     * Get the trip mask for a given date and query config.
      * <p>
-     * This method should return a map of route ids to trip masks for the given date.
+     * This method should return a map of route ids to trip masks for the given date and query config.
```
This is the beauty of this concept, there is no effect on performance or cache size. Of course in a productive system the available memory and cache size should be large holding many more raptor stop time instances than our default 5. And with LRU the more standard and common requests will always be cached. 

Also having an additional layer that manages which request goes to which instance of a cluster of raptor services to keep cache sizes small would be an option.

clukas1 commented:
@munterfi addressed most of your comments but struggling with one aspect.

> RoutingInfo vs. ScheduleInfo: Not consistent where this should live (ScheduleInfomrationService vs. ConnectionRoutingService) between the layers service and app.

Agree, implemented this and moved the Schedule Info to the ScheduleController. However after doing so, I realized that there is actually more to it and this is why I initially had the RouterInfo. Even if the Schedule holds information about Accessibility, Bikes and Travel Modes (which is probably always the case) it is not guaranteed that the Router used supports filtering routing solutions by it.

After this I've started implementing an additional RouterInfo logic, with the idea that the RaptorAlgorithm interface with have a getRouterInfo() method with these infos. Then again I realized that I couldn't just say that our "fancy" raptor implementation supports all these aspects because this again relies on the RaptorTripMaskProvider.

Solutions I see are:
- The service who knows Gtfs and Raptor references the Implementation of our Raptor and not the RaptorAlgorithmInterface in this way he holds the full information and since he also knows all the GtfsRaptor conversions -> MaskProvider, he can tell the app if the router supports query config values.
- The TripMaskProvider has to inform the raptor algorithm implementation if it supports query config parameters, this is then added to my initial plan that the RaptorAlgorithmInterface can return getRouterInfo() this is then passed up through the service up to the spring app.

> Why is there and option to have no travel modes? Do you know sources other than GTFS that are not providing this? In order to be generic we could keep it. But in our case this will always be supportsTravelModes=true

I didn't do research but there might be some strange format of schedules that does not include the vehicle type (for example mixed route types etc). Secondly and probably more relevant going back to the RouterInfo problem mentioned above, even if we have the information it is not clear if the service can filter connections to only allow certain travel modes.

> Configuration specific raptors and trip masks: Could you profile and benchmark to see how this is going to impact our memory footprint and performance?

Performance and cache size are only affected as you would expect. The addition of more options will result in more possible stop time arrays, thus fill up the cache quicker. As a result it's possible that the stop time arrays have to be recomputed more often (recomputation costs around ~1-1.5 s for a multi day raptor with 3 days to scan) but the performance of routing is not affected by the addition of more complex query configs. In a productive system, I would either or both use a cluster of raptor instances and manage my available stop time arrays in an additional layer before passing requests to the router or make sure that the raptor cache is really big >50 or more to ensure that the common requests never fall out of the cache (LRU).

munterfi reviewed (approved):
Thanks @clukas1, looks good to merge!
